### PR TITLE
add e2e test for error collector plugin.

### PR DESCRIPTION
**Pull request by jockech**
1 the transform logic is to emit input to error collector if id is less than 10.
2 move the `verify` logic to a common place so it can be used by `Joiner` and `ErrorCollector` features.
image

### CODE REVIEW

File: core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static io.cdap.plugin.common.stepsdesign.TestSetupHooks.gcsTargetBucketName;
```
Comment (itsankit-google) [inline core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java]: Is it needed? Can't we use `PluginPropertyUtils`?

File: core-plugins/src/e2e-test/features/errorcollector/ErrorCollector.feature [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    Then Open and capture pipeline preview logs
+    Then Verify the preview run status of pipeline in the logs is "succeeded"
+    Then Close the pipeline logs
+    Then Close the preview
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    # TODO: this line needs to change. where to put the code?
```
Comment (itsankit-google) [inline core-plugins/src/e2e-test/features/errorcollector/ErrorCollector.feature]: Is this `TODO` still relevant?

File: core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java (hunk 1 above)
Comment (jockech) [reply core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java]: The verification logic only works for `core-plugin`. is it okay to move to PluginPropertyUtils and PipelineSteps in e2e test repo?

Comment (jockech) [reply core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java]: after a second thought, the test case are specific to these plugins and we should maintain the logic in this repo.

File: core-plugins/src/e2e-test/features/errorcollector/ErrorCollector.feature (hunk 2 above)
Comment (jockech) [reply core-plugins/src/e2e-test/features/errorcollector/ErrorCollector.feature:72]: any suggestion on where to include this step? I think it should be put in `common` package, but each package only allows one java class in this repo.

File: core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java (hunk 1 above)
Comment (itsankit-google) [reply core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java]: No, I didn't meant to move the verification logic.

Lemme clarify, what I mean is why do we need to import `gcsTargetBucketName`? Can't we use `PluginPropertyUtils` to retrieve the property?

Comment (jockech) [reply core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java]: Thanks for clarifying it.

File: core-plugins/src/e2e-test/features/joiner/Joiner.feature [hunk 3]
```diff
@@ -75,5 +75,5 @@ Feature: Joiner analytics - Verify File source data transfer using Joiner analyt
     Then Wait till pipeline is in running state
     Then Open and capture logs
     Then Verify the pipeline status is "Succeeded"
-    Then Verify the CSV Output File matches the Expected Output File With Expected Partition
+    Then Verify the CSV Output File matches the Expected Output File: "joinerOutput" With Expected Partition: "expectedJoinerOutputPartitions"
```
Comment (rmstar) [inline core-plugins/src/e2e-test/features/joiner/Joiner.feature]: why are we making changes to joiner test here, seems unrelated to error collector?

**Review by rmstar — APPROVED** [review]

File: core-plugins/src/e2e-test/resources/pluginParameters.properties [hunk 4]
```diff
@@ -8,5 +8,19 @@ firstNameCsvFile=testdata/first_name.csv
...
 firstNameOutputSchema=[{"key":"id","value":"int"},{"key":"first_name","value":"string"}]
 lastNameOutputSchema=[{"key":"id","value":"int"},{"key":"last_name","value":"string"}]
+errorCollectorOutputSchema=[{"key":"id","value":"int"},{"key":"first_name","value":"string"},\
+  {"key":"msg","value":"string"},{"key":"code","value":"int"},{"key":"node","value":"string"}]
 joinerOutput=testdata/joiner_output.csv
-expectedPartitions=1
+errorCollectorOutput=testdata/error_collector_output.csv
+expectedJoinerOutputPartitions=1
+expectedErrorCollectorOutputPartitions=1
+transformScript=function transform(input, emitter, context) { \
```
Comment (itsankit-google) [inline core-plugins/src/e2e-test/resources/pluginParameters.properties]: nit: I think it is specfic to `javascript transform`, can we rename it to `javascriptTransformScript`?

File: core-plugins/src/e2e-test/resources/pluginParameters.properties [hunk 5]
```diff
@@ -8,5 +8,19 @@ firstNameCsvFile=testdata/first_name.csv
 lastNameCsvFile=testdata/last_name.csv
 firstNameOutputSchema=[{"key":"id","value":"int"},{"key":"first_name","value":"string"}]
 lastNameOutputSchema=[{"key":"id","value":"int"},{"key":"last_name","value":"string"}]
+errorCollectorOutputSchema=[{"key":"id","value":"int"},{"key":"first_name","value":"string"},\
+  {"key":"msg","value":"string"},{"key":"code","value":"int"},{"key":"node","value":"string"}]
 joinerOutput=testdata/joiner_output.csv
-expectedPartitions=1
+errorCollectorOutput=testdata/error_collector_output.csv
+expectedJoinerOutputPartitions=1
+expectedErrorCollectorOutputPartitions=1
```
Comment (itsankit-google) [inline core-plugins/src/e2e-test/resources/pluginParameters.properties]: Is error collector plugin also having some way to configure output partitions?

File: core-plugins/src/e2e-test/features/joiner/Joiner.feature (hunk 3 above)
Comment (jockech) [reply core-plugins/src/e2e-test/features/joiner/Joiner.feature]: I have move the `verifyOutput` logic to a same place under [CorePlugin.java](https://github.com/cdapio/hydrator-plugins/blob/38f130df787f004360d3069286faae812395516a/core-plugins/src/e2e-test/java/io/cdap/plugin/common/stepsdesign/CorePlugin.java) so it can be reused by `ErrorCollector`. Since the partition of `ErrorCollector` is always 1 but partition of `Joiner` can change, I introduced another parameter `partition` to the function, which then changes the original `Joiner` feature step.

File: core-plugins/src/e2e-test/resources/pluginParameters.properties (hunk 5 above)
Comment (jockech) [reply core-plugins/src/e2e-test/resources/pluginParameters.properties]: don't think so
image

File: core-plugins/src/e2e-test/resources/pluginParameters.properties (hunk 4 above)
Comment (jockech) [reply core-plugins/src/e2e-test/resources/pluginParameters.properties]: nice catch!

File: core-plugins/src/e2e-test/resources/pluginParameters.properties (hunk 5 above)
Comment (itsankit-google) [reply core-plugins/src/e2e-test/resources/pluginParameters.properties]: For such cases, can we pass expected partitions as `0` indicating we are not sure then it will combine the data from all the output files and sort it and verify?

Comment (jockech) [reply core-plugins/src/e2e-test/resources/pluginParameters.properties]: +1

**Review by itsankit-google — APPROVED** [review]
lgtm, pls squash commits before merge.
